datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -638,22 +638,26 @@ fn dictionary_coercion(
     preserve_dictionaries: bool,
 ) -> Option<DataType> {
     use arrow::datatypes::DataType::*;
+    let maybe_preserve_dictionaries =
+        |index_type: &Box<DataType>, value_type: DataType| -> DataType {
+            if preserve_dictionaries {
+                Dictionary(index_type.clone(), Box::new(value_type))
+            } else {
+                value_type
+            }
+        };
     match (lhs_type, rhs_type) {
         (
             Dictionary(_lhs_index_type, lhs_value_type),
             Dictionary(_rhs_index_type, rhs_value_type),
         ) => comparison_coercion(lhs_value_type, rhs_value_type),
```

Comment:
Is there any particular reason why this case doesn't  also preserve the dictionary type?

Reply:
I guess for comparison logic, preserving dict is not necessary, we just care about the value type.

Reply:
Probably because value type is enough for comparison

Reply:
Maybe we should not preserve dict for comparison overall 🤔 ?

Reply:
I think we should avoid `casting from value to dict for column`

Given the example, if we preserve dict, we still ends up casting column (utf8) to Dict (i32,utf8), but in this case, we can cast the const from i64 to utf8 and it is enough.

```
statement ok
create table test as values
  ('row1', arrow_cast(1, 'Utf8')),
  ('row2', arrow_cast(2, 'Utf8')),
  ('row3', arrow_cast(3, 'Utf8'))
;

# query using an string '1' which must be coerced into a dictionary string
query TT
SELECT * from test where column2 =  arrow_cast(2, 'Dictionary(Int32, Int64)');
----
row2 2

query TT
explain SELECT * from test where column2 = arrow_cast(2, 'Dictionary(Int32, Int64)');
----
logical_plan
01)Filter: CAST(test.column2 AS Dictionary(Int32, Utf8)) = Dictionary(Int32, Utf8("2"))
02)--TableScan: test projection=[column1, column2]
physical_plan
01)CoalesceBatchesExec: target_batch_size=8192
02)--FilterExec: CAST(column2@1 AS Dictionary(Int32, Utf8)) = 2
03)----MemoryExec: partitions=1, partition_sizes=[1]

statement ok
drop table test;
```


I think we should not preserve dict, but need specialization on  comparing dict vs non-dict case.

Reply:
> 02)--FilterExec: CAST(column2@1 AS Dictionary(Int32, Utf8)) = 2

Yes I agree that looks bad. It should be unwrapped. 

Thank you for that example 💯 

Maybe we could  extend https://github.com/apache/datafusion/blob/a5ce56831a9ec61e634d1c285c1b28d8c3891503/datafusion/optimizer/src/unwrap_cast_in_comparison.rs#L160  which handles `CAST(col, ..) = const` for other datatypes 🤔 

I can try to do so later this weekend or

Reply:
FWIW I tried the example from @jayzhan211  on main and it doesn't put a cast on the filter -- thus I agree this PR would be a regression if merged as is. I will dismiss my review

```
DataFusion CLI v37.1.0
> create table test as values
  ('row1', arrow_cast(1, 'Utf8')),
  ('row2', arrow_cast(2, 'Utf8')),
  ('row3', arrow_cast(3, 'Utf8'))
;
0 row(s) fetched.
Elapsed 0.050 seconds.

> explain SELECT * from test where column2 = arrow_cast(2, 'Dictionary(Int32, Int64)');

+---------------+---------------------------------------------------+
| plan_type     | plan                                              |
+---------------+---------------------------------------------------+
| logical_plan  | Filter: test.column2 = Utf8("2")                  |
|               |   TableScan: test projection=[column1, column2]   |
| physical_plan | CoalesceBatchesExec: target_batch_size=8192       |
|               |   FilterExec: column2@1 = 2                       |
|               |     MemoryExec: partitions=1, partition_sizes=[1] |
|               |                                                   |
+---------------+---------------------------------------------------+
2 row(s) fetched.
Elapsed 0.053 seconds.
```

Reply:
I agree with @jayzhan211 that we probably don't want to cast everything to dictionaries in the way that we are currently doing it, and what we really want is a way for the optimizer to avoid expensive casts of dictionary columns, and more generally to just avoid column casts in favor of casting constants and scalar expressions. 

I think what we have works fine for now and fixes performance issues we're seeing on dictionary columns, but should be improved for the general case in subsequent PRs that redesign the type coercion logic.